rust/src/mqtt/parser.rs

```diff
@@ -0,0 +1,631 @@
+                 MQTTUnsubackData {
+                   message_id: message_id,
+                   properties: properties,
+                   reason_codes: Some(reason_codes),
+                 }
+               )
+       ));
+
+#[inline]
+fn parse_disconnect(
+    input: &[u8],
+    mut remaining_len: usize,
+    protocol_version: u8,
+) -> IResult<&[u8], MQTTDisconnectData> {
+    if protocol_version < 5 {
+        return Ok((
+            input,
+            MQTTDisconnectData {
+                reason_code: None,
+                properties: None,
+            },
+        ));
+    }
+    if remaining_len == 0 {
+        // The Reason Code and Property Length can be omitted if the Reason
+        // Code is 0x00 (Normal disconnection) and there are no Properties.
+        // In this case the DISCONNECT has a Remaining Length of 0.
+        return Ok((
+            input,
+            MQTTDisconnectData {
+                reason_code: Some(0),
+                properties: None,
+            },
+        ));
+    }
+    match be_u8(input) {
+        Ok((rem, reason_code)) => {
+            remaining_len -= 1;
+            if remaining_len == 0 {
+                // no properties
+                return Ok((
+                    rem,
+                    MQTTDisconnectData {
+                        reason_code: Some(0),
+                        properties: None,
+                    },
+                ));
+            }
+            match parse_properties(rem, true) {
+                Ok((rem, properties)) => {
+                    return Ok((
+                        rem,
+                        MQTTDisconnectData {
+                            reason_code: Some(reason_code),
+                            properties: properties,
+                        },
+                    ));
+                }
+                Err(e) => return Err(e),
+            }
+        }
+        Err(e) => return Err(e),
+    }
+}
+
+named!(#[inline], pub parse_auth<MQTTAuthData>,
+       do_parse!(
+           reason_code: be_u8
+           >> properties: call!(parse_properties, true)
+           >>  (
+                 MQTTAuthData {
+                   reason_code: reason_code,
+                   properties: properties,
+                 }
+               )
+       ));
+
+pub fn parse_message(input: &[u8], protocol_version: u8) -> IResult<&[u8], MQTTMessage> {
+    // Parse the fixed header first. This is identical across versions and can
+    // be between 2 and 5 bytes long.
+    match parse_fixed_header(input) {
+        Ok((fullrem, header)) => {
+            let len = header.remaining_length as usize;
+            // This is the length of the fixed header that we need to skip
+            // before returning the remainder. It is the sum of the length
+            // of the flag byte (1) and the length of the message length
+            // varint.
+            let skiplen = 1 + varint_length(len);
+            let message_type = header.message_type;
```

Comment:
Can we add an assert that `skiplen == input.len() - fuller.len()` ?

Reply:
I guess. That would check whether `varint_length()` did the right thing...?

Reply:
I guess that I would rather compute `skiplen = input.len() - fullrem.len()` than using `varint_length`
So, not sure what the best is